google/cloud/forseti/scanner/scanners/kms_scanner.py

```diff
@@ -0,0 +1,136 @@
+            model_name (str): name of the data model
+            snapshot_timestamp (str): Timestamp, formatted as YYYYMMDDTHHMMSSZ.
+            rules (str): Fully-qualified path and filename of the rules file.
+        """
+        super(KMSScanner, self).__init__(
+            global_configs,
+            scanner_configs,
+            service_config,
+            model_name,
+            snapshot_timestamp,
+            rules)
+        self.rules_engine = kms_rules_engine.KMSRulesEngine(
+            rules_file_path=self.rules,
+            snapshot_timestamp=self.snapshot_timestamp)
+        self.rules_engine.build_rule_book(self.global_configs)
+
+    @staticmethod
+    def _flatten_violations(violations):
+        """Flatten RuleViolations into a dict for each RuleViolation member.
+
+        Args:
+            violations (list): The RuleViolations to flatten.
+
+        Yields:
+            dict: Iterator of RuleViolations as a dict per member.
+        """
+        for violation in violations:
+            yield {
+                'resource_id': violation.resource_id,
+                'resource_type': violation.resource_type,
+                'resource_name': violation.resource_name,
+                'full_name': violation.full_name,
+                'rule_index': violation.rule_index,
+                'rule_name': violation.rule_name,
+                'violation_type': violation.violation_type,
+                'violation_data': violation.violation_reason,
+                'resource_data': violation.resource_data
+            }
+
+    def _output_results(self, all_violations):
+        """Output results.
+
+        Args:
+            all_violations (list): All violations.
+        """
+        all_violations = list(self._flatten_violations(all_violations))
+        self._output_results_to_db(all_violations)
+
+    def _find_violations(self, keys):
+        """Find violations in the policies.
+
+        Args:
+            keys (list): CryptoKeys to find violations in.
+
+        Returns:
+            list: All violations.
+        """
+        all_violations = []
+        LOGGER.info('Finding crypto key rotation violations...')
+
+        for key in keys:
+            violations = self.rules_engine.find_violations(key)
+            LOGGER.debug(violations)
+            all_violations.extend(violations)
+
+        return all_violations
+
+    def _retrieve(self):
+        """Runs the data collection.
+
+        Returns:
+            list: CryptoKey objects.
+        Raises:
+            ValueError: if resources have an unexpected type.
+        """
+        keys = []
+
+        model_manager = self.service_config.model_manager
+        scoped_session, data_access = model_manager.get(self.model_name)
+        with scoped_session as session:
+            for key in data_access.scanner_iter(session, 'kms_cryptokey'):
+                if not key.parent_type_name.startswith('kms_keyring'):
```

Comment:
Is it possible to have any other values for parent than kms_keyring?

Reply:
No, added this to double check. Let me know if this needs to be removed.